Create Protobuf.Text, create `Protobuf.field_presence/2`

v0idpwn opened the pull request:
This PR introduces support for encoding Protobuf messages into the [textproto format](https://protobuf.dev/reference/protobuf/textformat-spec/). 

### Motivation  

The primary motivation for this feature is to improve the ability to inspect and visualize Protobuf messages, while still conforming to some standard. The inspect protocol default implementation can be painful to work with large or complex structures. Finding the relevant, non-default fields in a inspected struct is not intuitive and takes an unnecessary amount of time. A textproto representation provides a human-readable alternative that is much closer to the actual data transmitted over the wire (specially when using transformers). 

### This implementation

- **Encoding only**: Decoding was deliberately left out as it is less valuable compared to encoding, and has many caveats.
- **Emphasis on readability**: pretty-printed by default. We do a best effort for breaking lines, and opted for the most readable and elixir-like alternatives in the spec. Max line width ~~and tab size are~~ is configurable.
- **Limitations**:  
  - No support for extensions
  - No support for `Google.Protobuf.Any`.  
  - No type validation is performed. 

### Future Considerations  

I'm open to extending this implementation in the future to include:  
1. Adding support for extensions
2. Adding support for `Google.Protobuf.Any`.  
3. Implementing type validation.

v0idpwn committed:
Add Protobuf.Text

* Allows encoding to textproto

* Run conformance for text, fixes for conformance, add unknown fields support

Most of the conformance tests in the text suite are for decoding, and
some of them are for unknown fields encoding rules, so we only got value
out of five of them. Still, they led to significant improvements in the
implementation.

For one, I figured that the "outer message" **can't** use brackets ("{",
"}"). I thought they were optional and just not used in the examples. So
we went from:

```elixir
Protobuf.Text.encode(%MyMessage{a: 1, b: 2})
```

To:
```elixir
Protobuf.Text.encode(%MyMessage{a: 1, b: 2})
```

I also added unknown fields support, but wasn't able to pass all the
unknown fields tests, so I skipped them for now. The current version
does the job of printing them reasonably well, so I'm not too bothered
with it.

whatyouhide commented on lib/protobuf/text.ex:
```diff
@@ -0,0 +1,255 @@
+  According to https://protobuf.dev/reference/protobuf/textformat-spec/, without
+  extensions or `Google.Protobuf.Any` support.
+
+  Useful for inspecting/debugging protobuf-encoded data.
```
Everywhere else too please 😄

whatyouhide commented on lib/protobuf/text.ex:
```diff
@@ -0,0 +1,255 @@
+  Doesn't perform any validations. If input data is invalid, it produces
+  undecodable output.
+  """
+  def encode(struct, opts \\ []) do
```
Spec plz

whatyouhide commented on lib/protobuf/text.ex:
```diff
@@ -0,0 +1,255 @@
+              []
+
+            _ ->
+              raise "Invalid value for oneof `#{inspect(name)}`: #{inspect(value)}"
```
```suggestion
              raise "invalid value for oneof `#{inspect(name)}`: #{inspect(value)}"
```

whatyouhide reviewed (commented):
Should we look at using `Inspect.Algebra` to simplify the line splitting? Not sure what the spec of the format is and I don't have the bandwidth to go read it now, but it seems flexible enough that we could break lines in a few different places?

v0idpwn commented:
Pushed a new version using `Inspect.Algebra`. It indeed simplified code a lot, thanks for the suggestion. 

Some limitations/changes:
- I wasn't able to have different separators on `container_doc` for single line and multi-line. I'm sure it's achievable if I use the lower level functions, but I think having a `,` after every line isn't that bad (it's optional in the spec)
- I wasn't able to define `:pad_size` as I was in the hand-crafted implementation, so it's a fixed 2. 

I'll address the other comments over the week :)

whatyouhide commented:
@v0idpwn OK sounds good, lmk when this is gonna be ready for re-review. 🙃

v0idpwn commented on lib/protobuf/text.ex:
```diff
@@ -0,0 +1,155 @@
+  defp skip_field?(:proto3, "", %FieldProps{oneof: nil}), do: true
+  defp skip_field?(:proto3, false, %FieldProps{oneof: nil}), do: true
+
+  # This is actually new. Should it be ported to Protobuf.Encoder?
+  defp skip_field?(:proto3, value, %FieldProps{type: {:enum, enum_mod}, oneof: nil}) do
+    enum_props = enum_mod.__message_props__()
+    [first_tag | _] = enum_props.ordered_tags
+    %{name_atom: name_atom, fnum: fnum} = enum_props.field_props[first_tag]
+    value == name_atom or value == fnum
+  end
```
I tried this on Protobuf.Encoder and it broke conformance tests. Despite the fact that in the docs it seems like this is the expected behaviour there as well (just as we skip `0` for non-optional integers, I'd expect it to skip the default enum value...)

v0idpwn replied:
I'll investigate this further

v0idpwn commented:
I think it's okay to be reviewed, @whatyouhide :)

v0idpwn commented:
Actually, I found that there are conformance tests for text encoding. I'll work on them.

v0idpwn committed:
Bump google protos version

v0idpwn committed:
Create Protobuf.Presence module (#403)

* Create Protobuf.Presence module

Unifies presence tracking between encoders. Provides a public API for checking field presence

* Add test for messages, fix proto3

* Improve docs

v0idpwn committed:
Move the `field_presence` function to the `Protobuf` module

Also fix examples, and add an extra example about repeated

v0idpwn committed:
Add @doc since to Protobuf.Text

v0idpwn commented on test/protobuf/text_test.exs:
```diff
@@ -0,0 +1,150 @@
+      })
+
+    assert result == """
+           a: 1
+           e: {
+             a: 1,
```
I'm not very happy with this inconsistency: notice that the "outside" entries don't have `,`, but the nested ones do. Tried to make it match but failed. 

This is fine per the spec.

v0idpwn commented:
@whatyouhide I think it is. I did a rebase, removed stuff that was already merged (mostly related to CI), and tried to leave commits more optimized for a non-squash merge given the large scope of this PR (sorry about that, won't do it again 💀).

I did the google bump here because I wanted to ensure we had the latest conformance tests for text as well. I can move it to a separate PR if you wish.

v0idpwn committed:
Doc improvement in Protobuf.field_presence/2

v0idpwn committed:
Consistent map handling

Maps should be considered present when used for messages

v0idpwn committed:
format

whatyouhide commented:
@v0idpwn can we split up `Protobuf.field_presence/2` into its own PR?

v0idpwn commented:
Will do. Will split the google bump as well.
